src/io/flutter/samples/FlutterSampleNotificationProvider.java

```diff
@@ -5,65 +5,155 @@
-  @NotNull private final Project project;
+  @NotNull final Project project;
 
   public FlutterSampleNotificationProvider(@NotNull Project project) {
     this.project = project;
   }
 
   @NotNull
   @Override
-  public Key<JPanel> getKey() {
+  public Key<EditorNotificationPanel> getKey() {
     return KEY;
   }
 
   @Nullable
   @Override
-  public JPanel createNotificationPanel(@NotNull VirtualFile file, @NotNull FileEditor editor) {
-    final List<FlutterSample> samples = getSamplesForFile(file);
-    return !samples.isEmpty() ? new FlutterSampleActionsPanel(samples, project) : null;
-  }
-
-  private FlutterSdk getSdk() {
-    return FlutterSdk.getFlutterSdk(project);
-  }
+  public EditorNotificationPanel createNotificationPanel(
+    @NotNull VirtualFile file, @NotNull FileEditor fileEditor, @NotNull Project project) {
+    if (!(fileEditor instanceof TextEditor)) {
+      return null;
+    }
 
-  private List<FlutterSample> getSamplesForFile(@NotNull VirtualFile file) {
-    final FlutterSdk sdk = getSdk();
+    final FlutterSdk sdk = FlutterSdk.getFlutterSdk(project);
     if (sdk == null) {
-      return Collections.emptyList();
+      return null;
     }
 
+    final String flutterPackagePath = FileUtil.normalize(sdk.getHomePath()) + "/packages/flutter/lib/src/";
     final String filePath = FileUtil.normalize(file.getPath());
-    final String pathSuffix = FileUtil.normalize(sdk.getHomePath()) + "/packages/flutter/";
 
+    // Only show for files in the flutter sdk.
+    if (!filePath.startsWith(flutterPackagePath)) {
+      return null;
+    }
+
+    final TextEditor textEditor = (TextEditor)fileEditor;
+    final Editor editor = textEditor.getEditor();
+    final Document document = editor.getDocument();
+
+    final PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(document);
+    if (psiFile == null || !psiFile.isValid()) {
+      return null;
+    }
+
+    // Run the code to query the document in a read action.
+    final List<FlutterSample> samples = ApplicationManager.getApplication().
+      runReadAction((Computable<List<FlutterSample>>)() -> {
+        //noinspection CodeBlock2Expr
+        return getSamplesFromDoc(flutterPackagePath, document, filePath);
+      });
+
+    return samples.isEmpty() ? null : new FlutterSampleActionsPanel(samples);
+  }
+
+  private List<FlutterSample> getSamplesFromDoc(String flutterPackagePath, Document document, String filePath) {
     final List<FlutterSample> samples = new ArrayList<>();
-    for (FlutterSample sample : sdk.getSamples()) {
-      final String samplePath = pathSuffix + sample.getSourcePath();
-      if (filePath.equals(samplePath)) {
-        samples.add(sample);
+
+    // Find all candidate class definitions.
+    final PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(document);
+    assert (psiFile != null);
+
+    final List<DartClass> classDeclartions = new ArrayList<>();
+
+    psiFile.acceptChildren(new PsiElementVisitor() {
```

Comment:
Consider using, `PsiTreeUtil`, perhaps `getChildrenOfType`

Reply:
Cool, that's the util I was looking for; switched to `PsiTreeUtil`.